src/mon/FSCommands.cc

```diff
@@ -436,18 +437,38 @@ class SetHandler : public FileSystemCommandHandler
         return r;
       }
 
+      ss << fs->mds_map.get_fs_name();
+
       fsmap.modify_filesystem(
           fs->fscid,
           [is_down](std::shared_ptr<Filesystem> fs)
       {
-        if (is_down) {
-          fs->mds_map.set_flag(CEPH_MDSMAP_DOWN);
-        } else {
-          fs->mds_map.clear_flag(CEPH_MDSMAP_DOWN);
-        }
+	if (is_down) {
+	  fs->mds_map.set_flag(CEPH_MDSMAP_DOWN);
+	  fs->mds_map.set_old_max_mds();
+	  fs->mds_map.set_max_mds(0);
+	} else {
+	  mds_rank_t oldmax = fs->mds_map.get_old_max_mds();
+	  fs->mds_map.clear_flag(CEPH_MDSMAP_DOWN);
+	  fs->mds_map.set_max_mds(oldmax ? oldmax : 1);
+	}
       });
 
-      ss << "marked " << (is_down ? "down" : "up");
+      if (is_down) {
```

Comment:
I'm on the fence about this one. On the one hand, failing them would be much more efficient. On the other hand, it would be nice if a clean shutdown left the system in a clean state (i.e., without dirty journals laying around) that does not commit the user to restoring their previous state exactly.